File: src/main/java/com/euc/grantfitassessment/eval/DriftExperimentMain.java
package com.euc.grantfitassessment.eval;

import com.euc.core.EucDefinition;
import com.euc.core.EucLoader;
import com.euc.grantfitassessment.AlternateAlignmentPromptReasoner;
import com.euc.grantfitassessment.LlmFitReasoner;

import java.io.File;
import java.time.Instant;
import java.util.ArrayList;
import java.util.List;

/**
 * Runs the Week 5 drift experiment (docs/proposal.md Section 7) against a
 * live model: a baseline LlmFitReasoner vs. a prompt-variant reasoner
 * (AlternateAlignmentPromptReasoner, expected to alter behavior) and,
 * optionally, a model-swap variant if LLM_MODEL_VARIANT is set. Requires
 * ANTHROPIC_API_KEY — see README Build & Run.
 *
 * Run from the project root:
 *   mvn exec:java -Dexec.mainClass="com.euc.grantfitassessment.eval.DriftExperimentMain"
 *
 * Writes a JSON report to eval/grant-fit-assessment/results/ and prints a
 * summary. See DriftExperimentRunnerTest for an offline proof of the same
 * mechanics using a fake FitReasoner (no API key needed).
 */
public class DriftExperimentMain {

    private static final String DATASET_PATH = "eval/grant-fit-assessment/dataset/test-cases.json";
    private static final String RESULTS_DIR = "eval/grant-fit-assessment/results";

    public static void main(String[] args) {
        EucDefinition euc = EucLoader.loadGrantFitAssessment();
        List<TestCase> dataset = new TestCaseDataset().loadFromFile(new File(DATASET_PATH));

        String baselineModel = System.getenv().getOrDefault("LLM_MODEL", "claude-sonnet-4-6");
        FitReasonerVariant baseline = new FitReasonerVariant(
                "baseline:" + baselineModel, new LlmFitReasoner(baselineModel), false);

        List<FitReasonerVariant> candidates = new ArrayList<>();
        candidates.add(new FitReasonerVariant(
                "prompt-variant:loosened-alignment-instructions",
                new AlternateAlignmentPromptReasoner(baselineModel),
                true));

        String alternateModel = System.getenv("LLM_MODEL_VARIANT");
        if (alternateModel != null && !alternateModel.isBlank()) {
            candidates.add(new FitReasonerVariant(
                    "model-variant:" + alternateModel, new LlmFitReasoner(alternateModel), true));
        }

        DriftExperimentRunner runner = new DriftExperimentRunner(euc, dataset);
        DriftExperimentReport report = runner.run(baseline, candidates);

        System.out.println(report.summary());

        File outputFile = new File(RESULTS_DIR + "/drift-experiment-"
                + Instant.now().toString().replace(":", "-") + ".json");
        DriftExperimentReportWriter.writeJson(report, outputFile);
        System.out.println("Wrote " + outputFile);
    }
}

File: src/euc/grantfitassessment/eval/drift_experiment_main.py
"""Runs a controlled-change experiment: a baseline reasoner plus one or more
candidate variants, scored against the same dataset, to see whether the
mapped evaluators can tell a safe implementation change from a business
regression. Requires ANTHROPIC_API_KEY (used transitively by LlmFitReasoner)."""

from __future__ import annotations

import os
from datetime import datetime, timezone

from euc.core.loader import load_grant_fit_assessment
from euc.grantfitassessment.eval import dataset as dataset_module
from euc.grantfitassessment.eval.drift_experiment_report_writer import write_json
from euc.grantfitassessment.eval.drift_experiment_runner import DriftExperimentRunner
from euc.grantfitassessment.eval.fit_reasoner_variant import FitReasonerVariant
from euc.grantfitassessment.reasoner import AlternateAlignmentPromptReasoner, LlmFitReasoner

RESULTS_DIR = "eval/grant-fit-assessment/results"


def main() -> None:
    euc = load_grant_fit_assessment()
    dataset = dataset_module.load_from_file(dataset_module.DATASET_PATH)

    baseline_model = os.environ.get("LLM_MODEL", "claude-sonnet-4-6")
    baseline = FitReasonerVariant(
        label=f"baseline:{baseline_model}",
        reasoner=LlmFitReasoner(baseline_model),
        expected_to_alter_behavior=False,
    )

    candidates = [
        FitReasonerVariant(
            label="prompt-variant:loosened-alignment-instructions",
            reasoner=AlternateAlignmentPromptReasoner(baseline_model),
            expected_to_alter_behavior=True,
        )
    ]

    alternate_model = os.environ.get("LLM_MODEL_VARIANT")
    if alternate_model and alternate_model.strip():
        candidates.append(
            FitReasonerVariant(
                label=f"model-variant:{alternate_model}",
                reasoner=LlmFitReasoner(alternate_model),
                expected_to_alter_behavior=True,
            )
        )

    runner = DriftExperimentRunner(euc, dataset)
    report = runner.run(baseline, candidates)
    print(report.summary())

    timestamp = datetime.now(timezone.utc).isoformat().replace("+00:00", "Z").replace(":", "-")
    output_file = f"{RESULTS_DIR}/drift-experiment-{timestamp}.json"
    write_json(report, output_file)
    print(f"Wrote {output_file}")


if __name__ == "__main__":
    main()
